src/SubscriberBase.h

```diff
@@ -83,6 +83,15 @@ class SubscriberBaseT : public Subscriber<T>,
       bool startExecutor = true)
       : ExecutorBase(executor, startExecutor), cancelled_(false) {}
 
+  virtual ~SubscriberBaseT() {
+    if (!cancelled_ && originalSubscription_) {
```

Comment:
At first I thought the destructor is fine, but as I am thinking about how did you get into that I see a few issues.

If you called onSubscribe then you have to call a terminating on{Complete,Error}. If you didn't that is a broken contract and a bug. If you would call the terminating signal, you wouldn't have a valid pointer in the destructor.

From the threading correctness I am not sure the code is thread safe. We know that reading the pointer value or modifying the pointer is not thread safe. The destructor can be executed on any thread, but originalSubscription_ can be safely modified only in runInExecutor (like in the other methods). Since there could be 2 threads modifying the pointer close enough in time and there is no memory barrier (or lock) you might be reading stale value from processor cache. The only safe way to do this automatic cancellation is passing thisPtr inside the lambda, which makes no sense since you are in destructor and the instance is being destroyed.
I can't help it, but there is probably no safe way of doing it.

I would suggest that the only safe thing we can do in the destructor is (D)CHECK(! originalSubscription_); to make verify the user terminated the subscription..

What do you think?